HW-2 Model - Repository Pattern 구현

becooni opened the pull request.

becooni committed:
Update 전일대비 color Model에 저장

becooni committed:
Update repository pattern 적용

becooni committed:
Update Ticker 인자 이름 추가:

becooni committed:
Merge branch 'class04/junghum21' of https://github.com/StudyFork/GoogryAndroidArchitectureStudy into class04/junghum21

BeokBeok commented on junghum21/ArchStudy/app/src/main/java/com/jake/archstudy/data/source/UpbitRemoteDataSource.kt:
```diff
@@ -0,0 +1,57 @@
+    ) {
+        upbitService.getMarketAll()
+            .enqueue(object : Callback<List<MarketResponse>?> {
+                override fun onFailure(call: Call<List<MarketResponse>?>, t: Throwable) {
```
onResponse와 동일하게 파라미터를 정리해주세요.
```suggestion
                override fun onFailure(
                        call: Call<List<MarketResponse>?>,
                        t: Throwable
                 ) {
```

becooni replied:
https://github.com/StudyFork/GoogryAndroidArchitectureStudy/pull/188/commits/2c940ccc4f5e47087980f273660c262bf0859b1b
반영완료

BeokBeok commented on junghum21/ArchStudy/app/src/main/java/com/jake/archstudy/data/source/UpbitRepository.kt:
```diff
@@ -2,17 +2,22 @@ package com.jake.archstudy.data.source
+    override fun getMarketAll(
+        success: (List<MarketResponse>) -> Unit,
+        failure: (Throwable) -> Unit
+    ) {
```
한 줄로 표현이 가능합니다.
```suggestion
    = upbitRemoteDataSource.getMarketAll(
            success,
            failure
       )
```

becooni replied:
리턴하지 않고 콜백 함수만 전달해주기 떄문에 이렇게 작성했습니다.

BeokBeok replied:
리턴 값이 없다면, 기본값은 Unit 입니다.
해서 한 줄로 표현이 가능합니다.

becooni replied:
무언가를 return을 한다는 표현이므로 적절하지 않다고 생각이 드네요.

혼동을 줄수 있는 부분이라고 생각합니다.

코틀린 코딩컨벤션에는 Unit리턴시 =를 쓰지 말라고 한것은 아니지만,

샘플 코드에는 함수 바디 블럭을 유지하고 있습니다.

```kotlin
fun foo() { // ": Unit" is omitted here

}
```

> 코틀린 코딩 컨벤션 (Unit 리턴)
> https://kotlinlang.org/docs/reference/coding-conventions.html#unit

BeokBeok replied:
피드백 감사합니다 ^~^

BeokBeok commented on junghum21/ArchStudy/app/src/main/java/com/jake/archstudy/ui/MainActivity.kt:
```diff
@@ -29,31 +26,17 @@ class MainActivity : AppCompatActivity() {
+            val markets = response.groupBy { it.market.substringBefore("-") }
+                .map { map ->
+                    val title = map.key
+                    val markets = map.value.joinToString(separator = ",") {
```
joinToString 함수는 default separator가 콤마 이므로
파라미터를 생략해도 됩니다.

becooni replied:
https://github.com/StudyFork/GoogryAndroidArchitectureStudy/pull/188/commits/713e5bbd9c16a438865089d620a6f1c98909a246

반영완료

BeokBeok commented on junghum21/ArchStudy/app/src/main/java/com/jake/archstudy/ui/MainActivity.kt:
```diff
@@ -29,31 +26,17 @@ class MainActivity : AppCompatActivity() {
-                            initViewPager(markets)
-                        }
+        repository.getMarketAll({ response ->
+            val markets = response.groupBy { it.market.substringBefore("-") }
```
원소가 많을 경우, 원본 컬렉션을 시퀀스로 변환하여 처리하는 것을 권장합니다.

becooni replied:
https://umbum.tistory.com/599

컬렉션 성능에 관련된 것이군요.
한 수 배웠습니다. 감사합니다 :)

반영완료

https://github.com/StudyFork/GoogryAndroidArchitectureStudy/pull/188/commits/6f06cd03046e728c69a11759e01e5c8552180e2e

BeokBeok commented on junghum21/ArchStudy/app/src/main/java/com/jake/archstudy/ui/MainActivity.kt:
```diff
@@ -29,31 +26,17 @@ class MainActivity : AppCompatActivity() {
                 }
-            })
+            initViewPager(markets)
+        }, {})
```
가급적이면 에러 처리를 해주세요

becooni replied:
https://github.com/StudyFork/GoogryAndroidArchitectureStudy/pull/188/commits/66c411a539a4faa1fa52bcbb197c693edf3711cc

반영완료

BeokBeok reviewed (changes requested):
안녕하세요. 스터디 2기 강현석이라고 합니다.
부족하지만 잘 부탁 드립니다.

---

커밋 시, 앞에 "HW-주차" 를 붙여주세요.

becooni committed:
HW-2 onFailure 파라미터 개행

becooni committed:
HW-2 joinToString() 함수 seperator 생략

becooni committed:
HW-2 Collection Sequence 적용

Sequence로 변환후 컬렉션 함수를 돌리면, 컬렉션 함수 결과를 중간 계산 없이 최종 결과에만 적용한다.
https://umbum.tistory.com/599

becooni committed:
HW-2 network error 처리

haanbin commented on junghum21/ArchStudy/app/src/main/java/com/jake/archstudy/ui/TickersFragment.kt:
```diff
@@ -8,20 +8,17 @@ import androidx.recyclerview.widget.RecyclerView
     private val marketName by lazy { arguments?.getString(ARGS_MARKET_NAME) }
 
-    private val repository = UpbitRepository(ApiUtil.getUpbitService())
+    private val repository = UpbitRepository(UpbitRemoteDataSource(ApiUtil.getUpbitService()))
```
싱글톤으로 사용하는건 어떻게 생각하시나요 ?

becooni replied:
https://github.com/StudyFork/GoogryAndroidArchitectureStudy/pull/188/commits/ef1e07180b3fc95d70f308ed42dfc1838184e416

반영완료

haanbin replied:
수고하셨습니다!

haanbin reviewed (changes requested):
잘부탁드립니다!

becooni committed:
HW-2 toast ext 개선

1) 파라미터 개행
2) Fragment ext 함수는 Activity ext 함수를 호출하도록 변경

BeokBeok reviewed (dismissed):
LGTM:ok_hand:

becooni committed:
HW-2 repositry 싱글톤으로 구현

sjjeong reviewed (approved):
LGTM ✊ ✌️ 👋
